buildSrc/src/main/java/gradlebuild/JniNokeePlugin.java

```diff
@@ -0,0 +1,213 @@
+    }
+
+    private static String toVariantName(TargetMachine targetMachine) {
+        return targetMachine.getOperatingSystemFamily().getName() + "-" + targetMachine.getArchitecture().getName();
+    }
+
+    private void configureMainLibrary(Project project) {
+        JavaNativeInterfaceLibrary library = project.getExtensions().getByType(JavaNativeInterfaceLibrary.class);
+        registerWindowsDistributionDimension(library);
+        library.getTargetMachines().set(supportedMachines(library.getMachines()));
+        library.getTasks().configureEach(CppCompile.class, task -> {
+            task.getCompilerArgs().addAll(task.getTargetPlatform().map(targetPlatform -> {
+                OperatingSystem targetOs = targetPlatform.getOperatingSystem();
+                if (targetOs.isMacOsX()) {
+                    return ImmutableList.of("-mmacosx-version-min=10.9");
+                } else if (targetOs.isLinux()) {
+                    return ImmutableList.of("-D_FILE_OFFSET_BITS=64");
+                } else {
+                    return ImmutableList.of(); // do nothing
+                }
+            }));
+        });
+        library.getTasks().configureEach(LinkSharedLibrary.class, task -> {
+            task.getLinkerArgs().addAll(task.getTargetPlatform().map(targetPlatform -> {
+                OperatingSystem targetOs = targetPlatform.getOperatingSystem();
+                if (targetOs.isMacOsX()) {
+                    return ImmutableList.of(
+                        "-mmacosx-version-min=10.9",
+                        "-framework", "CoreServices");
+                } else if (targetOs.isWindows()) {
+                    return ImmutableList.of("Shlwapi.lib", "Advapi32.lib");
+                } else {
+                    return ImmutableList.of(); // do nothing
+                }
+            }));
+        });
+        library.getVariants().configureEach(variant -> {
+            if (variant.getBuildVariant().hasAxisOf(WindowsDistribution.WINDOWS_XP_OR_LOWER)) {
+                variant.getTasks().configureEach(CppCompile.class, task -> {
+                    task.getCompilerArgs().add("/DWINDOWS_MIN");
+                });
+            }
+        });
+        library.getVariants().configureEach(variant -> {
+            variant.getResourcePath().set(String.join("/",
+                project.getGroup().toString().replace('.', '/'),
+                "platform",
+                toVariantName(variant.getTargetMachine())));
+        });
+    }
+
+    private void registerWindowsDistributionDimension(JavaNativeInterfaceLibrary library) {
+        SetProperty<WindowsDistribution> newDimension = library.getDimensions().newAxis(WindowsDistribution.class, builder -> builder.onlyOn(library.getMachines().getWindows().getOperatingSystemFamily()));
+        newDimension.convention(ImmutableSet.copyOf(WindowsDistribution.values()));
+        ((ExtensionAware) library).getExtensions().add("targetWindowsDistributions", newDimension);
```

Comment:
🔍 See conversation in the `#ext-nokee` Slack channel around `ExtensionAware` guideline.